Add ERC: Purpose bound money (#128)
ethereum/ERCs: ERCS/erc-7291.md

---
eip: 7291
title: Purpose bound money
description: An interface extending ERC-1155 to implement purpose-bound money, a hybrid form of programmable payment and money.
author: Orchid-Dev (@proj-orchid-straitsx), Victor Liew (@alcedo), Wong Tse Jian (@wongtsejian), Chin Sin Ong (@chinsinong)
discussions-to: https://ethereum-magicians.org/t/eip-7291-purpose-bound-money/14973
status: Draft
type: Standards Track
category: ERC
created: 2023-06-24
requires: 165, 173, 1155
---

## Abstract

This proposal outlines a smart contract interface that builds upon the [ERC-1155](./eip-1155.md) standard to implement the purpose bound money (PBM) concept:

> - PBMs is comprised of a PBM wrapper and a digital money token that it wraps. A digital money token (e.g. stablecoins, central bank digital currency, tokenised bank deposits etc) serves as as a store of value (abbreviated as "sov"). Thus, a digital money token (also referred to as "sovToken") **SHOULD** be:

Reviewer: The uppercase keywords (like "SHOULD") create requirements, and should only be used in the specification section.

## Motivation

The establishment of this proposal seeks to forestall technology fragmentation and consequently a lack of interoperability. By making the PBM specification open, it gives new participants easy and free access to the pre-existing market standards, enabling interoperability across different platforms, wallets, payment systems and rails. This would lower cost of entry for new participants, foster a vibrant payment landscape and prevent the development of walled gardens and monopolies, ultimately leading to more efficient, affordable services and better user experiences.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Definitions

- **sovToken** - an ERC-20 or ERC-20 compatible digital currency (e.g. [ERC-777](./eip-777.md), [ERC-1363](./eip-1363.md)) serving as the store of value token (i.e. collateral backing the PBM Token).

> - **PBM Wrapper** - a smart ERC-1155 compliant smart contract, which wraps the sovToken, by specifying condition(s) that has/have to be met (referred to as PBM business logic in subsequent section of this proposal). The PBM wrapper can be designed to be modular in nature, with core, plug-ins and hooks components (see section on PBM Architecture for elaboration). The PBM wrapper smart contract, together with adopted hooks smart contracts verifies that condition(s) has/have been met before unwrapping the underlying sovToken.

Reviewer: I think this reads more easily?

```suggestion
- **PBM Wrapper** - an ERC-1155 compliant smart contract, which wraps the sovToken by specifying one or more conditions that have to be met (referred to as PBM business logic in subsequent section of this proposal). The PBM wrapper can be designed to be modular in nature, with core, plug-ins and hooks components (see section on PBM Architecture for elaboration). The PBM wrapper smart contract, together with adopted hooks smart contracts verifies that condition(s) has/have been met before unwrapping the underlying sovToken.
```

### Auditability

PBM Wrapper **SHOULD** provide the public easily accessible mechanism(s) to verify the smart contract logic for unwrapping a PBM. Such mechanisms could be leveraged by automated validation or asynchronous user verifications from transacting parties and/or whitelisted third parties attestations.

As the fulfilment of PBM conditions is likely to be subjected to audits to ensure trust amongst all transacting parties, the following evidence shall be documented to support audits:

- The interface/events emitted **SHOULD** allow a fine-grained recreation of the transaction history, token types and token balances
- The source code **SHOULD** be verified and formally published on a blockchain explorer.

### Fungibility

> A PBM Wrapper **SHOULD** be able to wrap multiple types of compatible Spot Tokens (e.g. the same PBM Wrapper should be able to wrap USDC and XSGD). Spot Tokens wrapped by the same PBM wrapper may or may not be fungible to one another. The standard does NOT mandate how an implementation must do this.

Reviewer: Is a Spot Token the same as a sovToken? I didn't see the term defined before here and I am unfamiliar with it.

Author: Yes, you are right, a Spot Token is the same as a sovToken. I have changed references to Spot Token to sovToken instead.

### PBM token details

> The ERC-1155 Multi Token Standard enables each token ID to correspond to a unique, configurable token type. All essential details facilitating the business or display logic for a specific PBM type **MUST** be defined for each token type. The mandatory fields for this purpose are outlined in the `struct PBMToken` (below). Future proposals may define additional, optional state variables as needed. Once a token detail has been defined, it **MUST** be immutable.

Reviewer: This section is a little vague on what the implementer actually has to do. It _sounds_ like the PBM Wrapper has to provide a way to access all the fields of the `PBMToken` struct, but it doesn't specify _how_ the implementer must do so. The way this is written, this could be a perfectly valid implementation:

```solidity
function details(uint256 tokenid) public returns (string, uint256, uint256, string, bool, bool, uint8, address, address, uint256, string, string) {
    // ...
}
```

I get the feeling that isn't what you meant.

Author: The PBM Wrapper (I.e. IPBMRC1) does not necessarily have to provide access to fields in the PBMToken struct. In this ERC, our suggested implementation is to have a function to retrieve the token details (for example, getTokenDetails(uint256 tokenid)) from a separate PBMRC1_TokenManager contract which can be shared by multiple PBM Wrapper contracts. 

As each token type can have a different PBMToken struct, we leave the exact implementation open to implementers. Thus, in the sample code, the getTokenDetails function returns the struct rather than individual fields. The implementer will have to extract individual fields from the PBM struct that is returned. 

Does this address your concerns?

Reviewer: It does, but this information should be in the proposal itself (if you haven't already added it.)

Example of token details:

```solidity
pragma solidity ^0.8.0;

> abstract contract IPBMRC1_TokenManager {

Reviewer: We generally prefer an interface over an abstract contract because the latter usually includes implementation details. Since this is an example, I think we can let it slide here.

        // OPTIONAL: Determines whether the PBM will be burned or revoked upon expiry, under certain predefined conditions, or at the owner's discretion.
        bool burnable;

        // OPTIONAL: Number of decimal places for the token.
        uint8 decimals;

        // OPTIONAL: The address of the creator of this PBM type on this smart contract. This field is optional because the creator is msg.sender by default.
        address creator;

        // OPTIONAL: The smart contract address of the sovToken.
        address tokenAddress;

        // OPTIONAL: The running balance of the PBM Token type that has been minted.
        uint256 totalSupply;

        // OPTIONAL: An ISO4217 three-character alphabetic code may be needed for the faceValue in multicurrency PBM use cases.
        string currencySymbol;

        // OPTIONAL: An abbreviation for the PBM token name may be assigned.
        string tokenSymbol;

        // Add other optional state variables below...
    }
}
```

An implementer has the option to define all token types upon PBM contract deployment. If needed, they can also expose an external function to create new PBM tokens at a later time.

> All token types created **SHOULD** emit a NewPBMTypeCreated event.

Reviewer: ```suggestion
All token types created **SHOULD** emit a `NewPBMTypeCreated` event.
```

    event TokenUnwrapForPBMBurn(address from, address to, uint256[] tokenIds, uint256[] amounts, address sovToken, uint256 sovTokenValue);

    /// Indicates the wrapping of an token into the PBM smart contract.
    /// @notice Emitted when underlying tokens are wrapped within the PBM smart contract.
    /// If there are multiple different underlying tokens involved in a single wrap operation, this event should be emitted for each underlying token.
    /// This event signifies the beginning of the PBM lifecycle, as tokens are now managed by the conditions within the PBM contract.
    /// @param from The address initiating the token wrapping process, and
    /// @param tokenIds An array containing the identifiers of the token types being wrapped.
    /// @param amounts An array containing the quantities of the corresponding wrapped tokens.
    /// @param sovToken The address of the underlying sovToken.
    /// @param sovTokenValue The amount of wrapped underlying sovTokens transferred.
    event TokenWrap(address from, uint256[] tokenIds, uint256[] amounts,address sovToken, uint256 sovTokenValue);
}

```

### Extensions

#### PBMRC1 - Token Receiver

> Smart contracts MUST implement all of the functions in the PBMRC1_TokenReceiver interface to subscribe to PBM unwrap callbacks.

Reviewer: ```suggestion
Smart contracts MUST implement all of the functions in the `PBMRC1_TokenReceiver` interface to subscribe to PBM unwrap callbacks.
```

When referring to interfaces, we prefer the use of backticks (\`).

        receiver smart contract supports type(PBMRC1_TokenReceiver).interfaceId

        @param _operator  The address which initiated the transfer (either the address which previously owned the token or the address authorised to make transfers on the owner's behalf) (i.e. msg.sender)
        @param _from      The address which previously owned the token
        @param _id        The ID of the token being unwrapped
        @param _value     The amount of tokens being transferred
        @param _data      Additional data with no specified format
        @return           `bytes4(keccak256("onPBMRC1BatchUnwrap(address,address,uint256,uint256,bytes)"))`
    */
    function onPBMRC1BatchUnwrap(address _operator, address _from, uint256[] calldata _ids, uint256[] calldata _values, bytes calldata _data) external returns(bytes4);
}

```

#### PBMRC2 - Non preloaded PBM Interface

The **Non Preloaded** PBM extension is OPTIONAL for compliant smart contracts. This allows contracts to bind an underlying sovToken to the PBM at a later date instead of during a minting process.

> Compliant contract **MUST** implement the following interface:

Reviewer: I'm assuming compliant here means compliant with the specific extension, and not with the whole document. Probably a good idea to disambiguate.
